Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -986,6 +986,46 @@ PhysicalParticleContainer::AddPlasma (int lev, RealBox part_realbox)
                     r = 1;
                 }
                 pcounts[index] = num_ppc*r;
+                // update pcount by checking which particles have non-zero density
+                int flag_pcount = 0;
+                for (int ip = 0; ip < num_ppc*r; ++ip) {
+                    amrex::Real dens1 = 0;
+                    amrex::Real dens2 = 0;
+                    amrex::Real dens3 = 0;
+                    amrex::Real dens4 = 0;
+                    amrex::Real dens5 = 0;
+                    amrex::Real dens6 = 0;
+                    amrex::Real dens7 = 0;
+                    amrex::Real dens8 = 0;
+                    // To ensure density parser does not encounter erroneous
+                    // arithmetic operation, such as log(0) or sqrt(-1)
+                    // check if cell-corner is within injection bounds
+                    if (inj_pos->insideBounds(lo.x, lo.y, lo.z))
+                        dens1 = inj_rho->getDensity(lo.x, lo.y, lo.z);
+                    if (inj_pos->insideBounds(lo.x, lo.y, hi.z))
+                        dens2 = inj_rho->getDensity(lo.x, lo.y, hi.z);
+                    if (inj_pos->insideBounds(lo.x, hi.y, lo.z))
+                        dens3 = inj_rho->getDensity(lo.x, hi.y, lo.z);
+                    if (inj_pos->insideBounds(hi.x, lo.y, lo.z))
+                        dens4 = inj_rho->getDensity(hi.x, lo.y, lo.z);
+                    if (inj_pos->insideBounds(lo.x, hi.y, hi.z))
+                        dens5 = inj_rho->getDensity(lo.x, hi.y, hi.z);
+                    if (inj_pos->insideBounds(hi.x, lo.y, hi.z))
+                        dens6 = inj_rho->getDensity(hi.x, lo.y, hi.z);
+                    if (inj_pos->insideBounds(hi.x, hi.y, lo.z))
+                        dens7 = inj_rho->getDensity(hi.x, hi.y, lo.z);
+                    if (inj_pos->insideBounds(hi.x, hi.y, hi.z))
+                        dens8 = inj_rho->getDensity(hi.x, hi.y, hi.z);
```

Comment:
@atmyers it was necessary to actually check insideBounds for all the points where we compute density
Otherwise I get an erroneous arithmetic operation especially for functions like `log(x)` where x is 0

Reply:
Maybe this could be made more compact by doing something like (unless you need the densities):
```
const auto xlim = GpuArray<Real, 2>{lo.x,hi.x};
const auto ylim = GpuArray<Real, 2>{lo.y,hi.y};
const auto zlim = GpuArray<Real, 2>{lo.z,hi.z};

const auto checker = [&](){
    for (const auto& x : xlim)
        for (const auto& y : ylim)
            for (const auto& z : zlim)
                if (inj_pos->insideBounds(x,y,z) &&  inj_rho->getDensity(x,y,z) > 0) return 1;
    return 0;
};

const auto flag_pcount = checker();
```

Reply:
oh thank you @lucafedeli88 !! 
I ll test this and then add it!